### PR TITLE
Handle running nested in a user namespace

### PR DESCRIPTION
**Pull request by hallyn**
Therefore check for that.  This is a step to being able to run docker/runc containers inside a user namespaced container.

No change in behavior when running in the initial user namespace.

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>

### CODE REVIEW

**Comment by hallyn** [comment]
Note there is no technical or security reason why these *could* not be allowed in a user namespace, but the kernel patch to change that was rejected by the kernel cgroups maintainer.

**Comment by mrunalp** [comment]
We had a related discussion in the OCI weekly call yesterday. I proposed separation of the devices cgroup configuration from the other device configuration. That way we can just exclude the device cgroup manipulation when nesting. However, this can also work before that lands. Your commit is missing the actual Godep vendored files, though. Thanks!

**Comment by hallyn** [comment]
On Thu, Jan 07, 2016 at 05:48:57PM -0800, Mrunal Patel wrote:
> Your commit is missing the actual Godep vendored files, though. Thanks!

yeah sorry wasn't sure what the right step is.

Should I 'go get' the code with runc/Godeps/_workspace as my GOPATH ?

**Comment by hqhq** [comment]
@hallyn runc uses godep, you can follow the instruction from: https://github.com/tools/godep#add-a-dependency

**Comment by dqminh** [comment]
@hallyn is it possible to move the required functions to another small-ish package ? Importing all the lxd/shared dependencies seem rather expensive for this small change.

**Comment by hallyn** [comment]
Hi - certainly possible :)  Ideally it would be defined in libcontainer and
reused in docker, which also needs to use this check at two places.

(see https://github.com/docker/docker/pull/19182)

I'll go ahead and update both requests with something shared that works.

**Comment by mrunalp** [comment]
@hallyn Thanks! Maybe in this file https://github.com/opencontainers/runc/blob/master/libcontainer/utils/utils.go

**Comment by hallyn** [comment]
@mrunalp 

I went with system instead of util, since it is linux-specific.  Should I move
it to utils instead?

**Comment by mrunalp** [comment]
@hallyn I am fine with system.

File: libcontainer/system/linux.go [hunk 1]
```diff
@@ -75,3 +78,29 @@ func Setctty() error {
...
+
+	buf := bufio.NewReader(file)
+	l, _, err := buf.ReadLine()
+	if err != nil {
+		return false
+	}
+
+	line := string(l)
+	var a, b, c int64
+	fmt.Sscanf(line, "%d %d %d", &a, &b, &c)
```
Comment (mrunalp) [inline libcontainer/system/linux.go]: Can we add a comment that this function assumes that uid_map has been written in the new user namespace (which is a fair assumption)?

Comment (hallyn) [inline libcontainer/system/linux.go]: On Fri, Jan 08, 2016 at 11:51:26AM -0800, Mrunal Patel wrote:
> > +func RunningInUserNS() bool {
> > +	file, err := os.Open("/proc/self/uid_map")
> > +	if err != nil {
> > +		return false
> > +	}
> > +	defer file.Close()
> > +
> > +	buf := bufio.NewReader(file)
> > +	l, _, err := buf.ReadLine()
> > +	if err != nil {
> > +		return false
> > +	}
> > +
> > +	line := string(l)
> > +	var a, b, c int64
> > +	fmt.Sscanf(line, "%d %d %d", &a, &b, &c)
> 
> Can we add a comment that this function assumes that uid_map has been written in the new user namespace (which is a fair assumption)?

I don't understand what you mean.  The file is a procfile, served by the
kernel.  If it's not there we assume the kernel doesn't support user
namespaces and return false at the top.

Happy to add a comment but not sure what you are referring to.

Comment (mrunalp) [inline libcontainer/system/linux.go]: I meant that we are checking for the default values in the uid_map file to check user namespace is in use. So the assumption is that new values that differ from the default values in that file have been written. Adding this is discretionary but I think might clarify the behavior. What do you think?

Comment (hallyn) [inline libcontainer/system/linux.go]: On Fri, Jan 08, 2016 at 12:47:24PM -0800, Mrunal Patel wrote:
> > +func RunningInUserNS() bool {
> > +	file, err := os.Open("/proc/self/uid_map")
> > +	if err != nil {
> > +		return false
> > +	}
> > +	defer file.Close()
> > +
> > +	buf := bufio.NewReader(file)
> > +	l, _, err := buf.ReadLine()
> > +	if err != nil {
> > +		return false
> > +	}
> > +
> > +	line := string(l)
> > +	var a, b, c int64
> > +	fmt.Sscanf(line, "%d %d %d", &a, &b, &c)
> 
> I meant that we are checking for the default values in the uid_map file to check user namespace is in use. So the assumption is that new values that differ from the default values in that file have been written. Adding this is discretionary but I think might clarify the behavior. What do you think?

Gotcha - will do

**Comment by hallyn** [comment]
Is travis being fragile, or am I just failing to see what i did wrong?

**Comment by mrunalp** [comment]
@hallyn Thanks! LGTM 
ping @crosbymichael @LK4D4

**Comment by dqminh** [comment]
LGTM

**Pull request merged**